### DISCUSSION TITLE
WiFI disconnects / stability problem when WPA3 is enabled (transition Mode from WPA2) within Mesh on AVM products

### DISCUSSION THREAD

**Discussion by tp1de** [emsesp/EMS-ESP32#1073] [General]
I am using a Mesh-Network based on AVM products. There are 3 active WiFi components:

1. Cable-Router far way from ems-esp (-90db) - Mesh master
2. old Router acting as Mesh access point - next room to ems-esp (-60 db)
3. a WiFi Repeater close to ems-esp (50 cm)  (- 38 db)

All WiFi attributes are set central on Mesh master are then identically set to all Mesh devices.
AVM is recommending to select a WPA2/WPA3 transition mode for WiFi security. (WPA2 and WPA3 enabled).

With this mode all my devices run stable but ems-esp does not. The ems-esp firmware is changing regularly the access points to be connected to. This is not the strongest one ! 
I could see  that quite often the connection is made to the far away master.

Being away from home two weeks ago I could not connect anymore to the WebUI - but MQTT connection to HA was still working.
After 6 hours another access point was selected and connection to WebUI was possible again.

When using the WPA2/WPA3 transition mode the ems-esp WiFi network scan does not show the right security attributes for the mesh repeaters - only for the master. I might believe that this causes the connection problems.

Selecting fixed WPA2 for WiFi security is recognized for all access points by ems-esp and connection seems to be stable now.
(even changing the access points to be connected to a couple of times per day !)

**Comment by proddy** [comment]
There's not a lot we can do here in EMS-ESP. We're using the core espressif wifi libraries. You could scan through the forums (https://github.com/espressif/esp-idf) and see if other people have similar issues? Or try the tec build as uses the latest ESP-IDF 5.0 library.

On my Unifi network there are settings to optimize for iot devices - maybe AVM has something similar? Or just fix the IP and AP for the EMS-ESP device?

Or use an ESP32 microcontroller by itself and write a small script that connects to the network to see if you can reproduce it easily.

**Reply by tp1de** [reply]
I will watch if fixing to WPA2 might solve disconnects even with the regular changes of access points.
If the disconnects disappear then it might be good to include a WPA3 warning in the documentation.

I just wanted to share my findings so far. 
The AVM firmware is changed actually and actual versions are different per device. iot specific settings are not available.
And yes not using Mesh and having different SSID's per access point would solve this as well - but I do not want ...

**Reply by proddy** [reply]
I'll do some research too. Thanks for reporting it. 
I didn't mean setting different SSIDs across the APs, but hardcoding EMS-ESP to one AP using its MAC address.

**Reply by proddy** [reply]
https://github.com/espressif/arduino-esp32/issues/6873
https://esp32.com/viewtopic.php?t=22786

Looks like WPA3 is only supported on the ESP32-S2 and S3 (not the BBQKees gateway boards). You could swap out the ESP32?

**Reply by tp1de** [reply]
> I didn't mean setting different SSIDs across the APs, but hardcoding EMS-ESP to one AP using its MAC address.

Couldn't this be done within the Scan Wifi Networks dialog where you select one access point anyhow and see the signal strength?

**Reply by MichaelDvP** [reply]
The latest espressif API supports WPA3: https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-guides/wifi-security.html

Afaik most mesh issues are caused by automatic channel changes of the AP. On channel change the connection is disconnected and reconnect is the next possible AP, maybe far away and low signal.

**Reply by tp1de** [reply]
![image](https://user-images.githubusercontent.com/36480251/220315803-dc884979-d4f4-47f3-9b58-6cd3e2ca348c.png)

**Reply by tp1de** [reply]
> Afaik most mesh issues are caused by automatic channel changes of the AP.

I know and therefore I use a fixed channel.

**Reply by proddy** [reply]
> The latest espressif API supports WPA3: https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-guides/wifi-security.html
> 
> Afaik most mesh issues are caused by automatic channel changes of the AP. On channel change the connection is disconnected and reconnect is the next possible AP, maybe far away and low signal.

but not the arduino which we use. We should perhaps fork our own version of the SDK like Tasmota and other projects do. https://docs.espressif.com/projects/arduino-esp32/en/latest/troubleshooting.html?highlight=wpa3#why-does-the-board-not-connect-to-wpa3-encrypted-wi-fi

**Reply by MichaelDvP** [reply]
Hmm, this is defined in the arduino: https://github.com/espressif/arduino-esp32/blob/0d84018d969309addacbcc3e3782c1fadc95fbc8/tools/sdk/esp32/dio_qspi/include/sdkconfig.h#L356
Maybe that's the cause for the much more flash the newer platform needs?

**Comment by tp1de** [comment]
WiFi is stable now, when I enable fixed WPA2 within my Mesh network. (WPA2/WPA3 transition mode disabled).
Nevertheless EMS-ESP connects to different access-points regularly. I have no glue why and if there is any short disconnect.
With the changes done for MQTT there are no reboots anymore.

![image](https://user-images.githubusercontent.com/36480251/221540860-4294c8ab-0908-46e4-ab11-3c93bb152f23.png)

![image](https://user-images.githubusercontent.com/36480251/221541841-c0a307e1-53fa-4d0d-9bac-29e0db40fbaa.png)

**Comment by tp1de** [comment]
I recognized that when EMS-ESP reboots (e.g. new firmware upload) WiFi is not automatically reconnected to the strongest access point.

The desired function would be, that on reconnect / reboot a WiFi network scan is done and the strongest access point is used with same SSID and password as stored. (same as selecting it manually but where password has to re-entered even if SSID stays the same).

**Reply by proddy** [reply]
it's the job of the router to assign a client to an access point and assign an IP. I'm not sure what EMS-ESP (being the client) can influence.

**Reply by tp1de** [reply]
With a WiFi network scan from WebUI I see all access points  with SSID and signal strength. In my case a couple of them with same SSID. There I can select one ems-esp to be connected. So functionality is already available.

.... YES you might be right. I selected a weaker access point and it doesn't work. So forget it.

**Reply by proddy** [reply]
probably by chance. There is only one line in EMS-ESP to connect to the wifi

```cpp
WiFi.begin(_state.ssid.c_str(), _state.password.c_str()); // attempt to connect to the network
```

which takes a SSID and a Password.

**Comment by tp1de** [comment]
I tested 3.6.0-dev.6 with WPA2/WPA3 transition mode enabled again. Fritz!Box Mesh Network with Cable Router, One access point connected by  LAN and a second one by Wifi in the heating room. Buderus Boiler connected by LAN to this and ems-esp connect by WLAN to this access point (50 cm distance).

The findings:
1. ems-connects not always to the nearest and strongest signal access point - this can't be influenced (as discussed before) 

2. with wpa2/3 transition mode within wifi network scan all 3 access points are shown but `Security: Unknown` 

3. when connected to an access point in 2 meters distance (1 wall in between) I recognized a lot of spikes in WiFi RSSI time graph:
![image](https://user-images.githubusercontent.com/36480251/224724147-ef48c20d-4462-4aa7-8ce0-4365283a514a.png)

4. These spikes are 3-4 times more often then by just using wpa2.

Conclusion I do not feel save with wpa3/3 transition mode and I will go back to wpa2.

**Reply by proddy** [reply]
that's a shame. Unfortunately nothing we can fix in EMS-ESP.

**Comment by caveman2003** [comment]
Hi tp1de, 

i have the same issue with AVM... but in addition to the mesh of AVM (2 devices) i have also 2 TP Link (Omada) devices. When i take the device in my working room (TP Link AP 4m distance) WebUI is running perfect, moving back to cellar where only FritzBox is visible (scan show only the box, 2m distance, no walls) WebUI is not reachable anymore... I have also WPA2/WPA3. 
And as you, i see on the ESP32 "security unknown". Changing to WPA2(CCMP) and scan again "security WPA2" is shown...

From my point of view this is realy strange because authentication to WLAN is done only on the Fritz not on the TP Link ...

Strange is that i have in the cellar an other ESP32 with Tasmota (+ an Pi Zero... but this is totaly different). The ESP32 with Tasmota haven't show any problems with WPA2/WPA3....

**Reply by tp1de** [reply]
Yes I still confirm WPA3 connection problems with AVM devices. I think this is not related to mesh but rather the WPA3 implementation.
Since I switched back to WPA2(CCMP) the WiFi connection is long-term stable.

**Reply by proddy** [reply]
I've running EMS-ESP on an S3 with the latest dev 3.6.x, on WPA3 for the last few days and haven't noticed any drops.

**Comment by tp1de** [comment]
I think the problem is related to AVM's FritzBox WPS2/WPA3 transition mode.

**Reply by MichaelDvP** [reply]
I've tested with AVM fritzbox 7530 and have no issues with WPA2/WPA3. But i don't have a mesh.

**Reply by tp1de** [reply]
I do have a FB 6591 with Vodafone Firmware 7.29 and two Mesh repeaters with different FW releases. This might give the problem with WPA3 ... but WPA2 is fine and stable for the meantime.
Do you already have 7.5x ?

**Reply by MichaelDvP** [reply]
> Do you already have 7.5x ?

V7.56 (release) and some "labor" before. 
Ems-esp shows in network scanner `Security unknown`, i'll add the modes WPA2_WAP3_PSK(7) and WPA3_PSK(6) for web-display. For the fritzbox is shows than WPA/WPA3, correctly identified.
![grafik](https://github.com/emsesp/EMS-ESP32/assets/59284019/b571d44a-549d-49c5-8b46-3549fac3f3d1)
And ems-esp connects with WPA3:
![grafik](https://github.com/emsesp/EMS-ESP32/assets/59284019/900da111-38bb-46bf-bb99-9681445f0802)

**Reply by MichaelDvP** [reply]
Seems the WPA3 is only stable if compiiled with the latest dev platform. The official esp32 build with older platform (due to memory) have issues with the  WPA3 mixed mode. The S3 build works.

**Reply by proddy** [reply]
On my to-do is still to figure out why the latest IDF v4.4.4 increases the Flash by 20%. I was going to try and use some of the analysis tools to compare the two to find which library is causing it and then find ways to disable it. I couldn't get the pio analysis to work so looking now into https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-guides/performance/size.html . Then we can all use [platform-espressif32](https://github.com/platformio/platform-espressif32) version v6.3.2
